### ISSUE TITLE
warning: queue buffer size exhausted, skipping scheduling of task..

### ISSUE THREAD

**Issue by mvisonneau** [mvisonneau/gitlab-ci-pipelines-exporter#327]
I'm having the below error messages in the exporter pod:
```
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"254307853","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"467127557","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"4117795749","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"986762252","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"1234608350","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"776945100","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"3467447372","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"2171559493","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"3561156129","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"131636616","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"1717533351","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"1935164250","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"1140558331","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"1880386309","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"4134022478","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"2157454054","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"3875675952","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"4075550179","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"1793384940","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"2495151196","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"1453467045","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"1228075757","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"2528858446","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"582813482","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"2552249403","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"3358890680","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"1277022775","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"3757704787","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"2827269068","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"75705386","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"4246894544","time":"2021-09-15T12:16:25Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullRefMetrics","task_unique_id":"820641972","time":"2021-09-15T12:16:25Z"}
```

Does this look famliar?

_Originally posted by @BilelYahia in https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/issues/308#issuecomment-919976453_

**Comment by mvisonneau** [comment]
:wave: hey @bilelyahia, this warning may occur when the exporter is attempting to schedule more than 1000+ jobs. The main root cause for this would be the combination of one or more configuration parameters:
- too short polling intervals 
- too low GitLab API rate limit
- too many projects, refs, environments or metrics

**Comment by vsimon** [comment]
Just a data point, I also notice messages like these too in the logs, running gitlab on gitlab.com:

```
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullEnvironmentsFromProject","task_unique_id":"2962900000","time":"2021-09-15T22:58:48Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullEnvironmentsFromProject","task_unique_id":"1725200000","time":"2021-09-15T22:58:48Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullEnvironmentsFromProject","task_unique_id":"2781700000","time":"2021-09-15T22:58:48Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullEnvironmentsFromProject","task_unique_id":"1327500000","time":"2021-09-15T22:58:48Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullEnvironmentsFromProject","task_unique_id":"1513400000","time":"2021-09-15T22:58:48Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullEnvironmentsFromProject","task_unique_id":"3394200000","time":"2021-09-15T22:58:48Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullEnvironmentsFromProject","task_unique_id":"3515200000","time":"2021-09-15T22:58:48Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullEnvironmentsFromProject","task_unique_id":"2899300000","time":"2021-09-15T22:58:48Z"}
{"level":"warning","msg":"queue buffer size exhausted, skipping scheduling of task..","task_type":"PullEnvironmentsFromProject","task_unique_id":"1613400000","time":"2021-09-15T22:58:48Z"}
```

as well as some of these

```
{"jobs-count":0,"level":"warning","msg":"found some ref jobs but did not manage to refresh all jobs which were in memory","not-found-jobs":"xxxxxxx","project-name":"xxxxxxx","ref":"xxxxxxx","time":"2021-09-15T22:58:54Z"}
{"jobs-count":0,"level":"warning","msg":"found some ref jobs but did not manage to refresh all jobs which were in memory","not-found-jobs":"xxxxxxx","project-name":"xxxxxxx","ref":"xxxxxxx","time":"2021-09-15T22:58:59Z"}
```

**Comment by vsimon** [comment]
`image: mvisonneau/gitlab-ci-pipelines-exporter:v0.5.2`

**Comment by perlun** [comment]
(What we did for now is to use webhooks instead, i.e. push the GitLab updates to the exporter process instead. We can then poll as well to be able to catch potentially missing data.)

**Comment by ocrlgv** [comment]
hey @perlun , how did you use the webhook? is that a system webhook or per project? how does the hook update this project instance as I thought it pulled everything.
I got 400+ projects so seems some settings would need to be increased but feel the slowness then makes job metrics useless if too slow.

**Comment by perlun** [comment]
@ocrlgv Sorry for the delay. We've set it up using a group-level webhook, since all (or a large majority) of our projects are located in a single group. This works good-enough for us. See https://github.com/mvisonneau/gitlab-ci-pipelines-exporter#push-based-implementation-leveraging-gitlab-webhooks for more details on how to configure this with the exporter.

**Comment by DerFels** [comment]
Any news on how to fix this issue? I just started the exporter on version 0.5.3 and immediately the log is full of
```
time="2022-03-08T15:10:36Z" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=3992715563
time="2022-03-08T15:10:36Z" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=4089987140
time="2022-03-08T15:10:36Z" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=1764227896
```

**Comment by brandoncamenisch** [comment]
Same here I have a fairly large GitLab setup using mostly the default config  for this exporter but targeting groups/sub-groups with group level webhooks and it's not indexing everything. Is there some configuration changes one can make to get it working or perhaps HA with multiple exporters attached to different groups. 

It's like it's hitting a limit and can't go beyond that. 😢

**Comment by oijkn** [comment]
Same here with large Gitlab setup.

**Comment by mvisonneau** [comment]
I am happy to parameterize this value. However, it will unlikely solve the implementation issues you are currently facing. If you exhaust this buffer size, it means that your configuration needs to be adjusted in order to fit the amount of data you are looking to pull with the current GitLab API rate limitation you are enforcing. So as I already said previously, you need to one or more of the following actions:

* increase polling intervals
* increase GitLab API rate limit
* reduce the amount of projects, refs, environments or metrics you are polling
* disable scheduled polling and leverage webhooks instead

**Comment by 41lx** [comment]
> I am happy to parameterize this value. However, it will unlikely solve the implementation issues you are currently facing. If you exhaust this buffer size, it means that your configuration needs to be adjusted in order to fit the amount of data you are looking to pull with the current GitLab API rate limitation you are enforcing. So as I already said previously, you need to one or more of the following actions:
> 
> * increase polling intervals
> * increase GitLab API rate limit
> * reduce the amount of projects, refs, environments or metrics you are polling
> * disable scheduled polling and leverage webhooks instead

Hi,

it seems that we are facing same limit using webhooks, using configuration:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: gitlab-exporter-webhook-config
  namespace: platform
data:
  gitlab-ci-pipelines-exporter.yml: |
    log:
      level: debug
      format: text # json
    gitlab:
      url: https://gitlab.xyz.com
      token: Y
    server:
      webhook:
        enabled: true
        secret_token: X
    wildcards:
      - {}
 ```

```
time="2022-12-01T15:02:22Z" level=debug msg="webhook request" ip-address="10.13.236.6:51164" user-agent=GitLab/15.5.3-ee
time="2022-12-01T15:02:22Z" level=info msg="received a pipeline webhook from GitLab for a ref, triggering metrics pull" project-name=elis/scheduler/merge-if-tested ref=master ref-kind=branch
time="2022-12-01T15:02:22Z" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=2812917000
time="2022-12-01T15:02:22Z" level=debug msg="throttled GitLab requests" for=936.543062ms
time="2022-12-01T15:02:23Z" level=debug msg="throttled GitLab requests" for=950.335305ms
time="2022-12-01T15:02:23Z" level=info msg="discovered new ref" project-name=mobile-analytics-library ref=g ref-kind=tag
time="2022-12-01T15:02:23Z" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=2499141481
time="2022-12-01T15:02:23Z" level=info msg="discovered new ref" project-name=mobile-analytics-library ref=master ref-kind=branch
time="2022-12-01T15:02:23Z" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=3198381950
time="2022-12-01T15:02:24Z" level=debug msg="throttled GitLab requests" for=947.75364ms
time="2022-12-01T15:02:25Z" level=debug msg="throttled GitLab requests" for=903.305734ms
time="2022-12-01T15:02:26Z" level=debug msg="throttled GitLab requests" for=956.267483ms
time="2022-12-01T15:02:26Z" level=info msg="discovered new ref" project-name=python-demo ref=master ref-kind=branch
time="2022-12-01T15:02:26Z" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=3099642792

**Comment by DurgeshBarve** [comment]
> Hi @mvisonneau, also running into this.
> 
> > this warning may occur when the exporter is attempting to schedule more than 1000+ jobs. The main root cause for this would be the combination of one or more configuration parameters:
> > ```
> > * too short polling intervals
> > * too low GitLab API rate limit
> > * too many projects, refs, environments or metrics
> > ```
> 
> I seem to get this when the exporter goes to pull metrics:
> 
> ```
> Oct 20 08:55:12 prometheus01 gitlab-ci-pipelines-exporter[69006]: time="2021-10-20T08:55:12+03:00" level=info msg="scheduling metrics pull" environments-count=0 refs-count=1128
> Oct 20 08:55:12 prometheus01 gitlab-ci-pipelines-exporter[69006]: time="2021-10-20T08:55:12+03:00" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=2695544300
> Oct 20 08:55:12 prometheus01 gitlab-ci-pipelines-exporter[69006]: time="2021-10-20T08:55:12+03:00" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=3433616333
> Oct 20 08:55:12 prometheus01 gitlab-ci-pipelines-exporter[69006]: time="2021-10-20T08:55:12+03:00" level=warning msg="queue buffer size exhausted, skipping scheduling of task.." task_type=PullRefMetrics task_unique_id=1136253962
> ```
> 
> I noted that this seems to be a constant in the code. Would it make sense to allow the buffer to be extended via the YAML config file? The default of 1000 could still be kept, but for environments where the number of projects/refs are too large I think a buffer size of 1000 will simply be inadequate sometimes. WDYT, would it make sense?
> 
> https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/blob/dbaed73c54f16b34640d46489def737690b5433d/pkg/controller/scheduler.go#L18
Do you  got any solution for the issue?

**Comment by RamyHuang** [comment]
How to set the parameter value

**Comment by mvisonneau** [comment]
in case you really need it, I've added a [new configuration parameter](https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/commit/90444f156508db4457a4025dcd7851a63ec5aaa8#diff-4194838c0fad9762c5a0e565dd61998e713f6710b58d5d8bd6bf19005ab589caR97-R106) (available from [v0.5.5](https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/releases/tag/v0.5.5))

**Comment by stewarca** [comment]
> 👋 hey @bilelyahia, this warning may occur when the exporter is attempting to schedule more than 1000+ jobs. The main root cause for this would be the combination of one or more configuration parameters:
> 
> * too short polling intervals
> * too low GitLab API rate limit
> * too many projects, refs, environments or metrics

@mvisonneau Can I ask what you mean by too short of polling intervals here?
